common/log/logger.go

```diff
@@ -151,6 +158,26 @@ func (lg *loggerImpl) WithTags(tags ...tag.Tag) Logger {
 	}
 }
 
+// DebugOn checks if debug level is on.
+// This is useful to avoid expensive debugging serializations etc. in production.
+// It caches the result for debugOnCheckInterval and checks again if the interval has passed.
+// Log level changes not reflected immediately because of the cache but it's acceptable.
+func (lg *loggerImpl) DebugOn() bool {
+	if time.Since(time.Unix(0, atomic.LoadInt64(&lg.debugOnCheckTimestampNanos))) < lg.debugCheckInterval {
+		return atomic.LoadInt32(&lg.debugOn) != 0
+	}
+
+	on := int32(0)
+	if lg.zapLogger.Check(zap.DebugLevel, "test") != nil {
+		on = 1
+	}
+
+	// no locking to avoid performance overhead. there's chance of redundant computation but it's acceptable.
+	atomic.StoreInt32(&lg.debugOn, on)
```

Comment:
tbh I wouldn't bother with this without a benchmark.  Zap already has `zap.Check` and it's already atomics-oriented in most code I've seen, and it's intended for this exact purpose

Reply:
We are already using `zapLogger.Check` above. However it's not free as it creates structs the `zapcore.Entry` and formats the message before doing the check which is what we want to avoid for expensive computations in hot path only to enrich debug logs.

Reply:
from chat elsewhere:
`func (lg *loggerImpl) DebugOn() bool {` is entirely reasonable and I'm fully in favor of it.  it has clear usefulness, clear semantics, etc.

it's more that this:
```go
lg.zapLogger.Check(zap.DebugLevel, "test")
```
is _faster_ than after adding the atomics.  just do that every time instead.

Reply:
Thanks for bench'ing this. It got auto-merged. I will address this in a follow up PR